internal/icinga2/client.go

```diff
@@ -0,0 +1,385 @@
+
+		// catchupErr might hold an error received from catchupEventCh, indicating another catch-up-phase run.
+		catchupErr error
+	)
+
+	// catchupReset resets all catchup variables to their initial empty state.
+	catchupReset := func() {
+		catchupEventCh, catchupCancel = nil, nil
+		catchupBuffer = make([]*event.Event, 0)
+		catchupCache = make(map[string]time.Time)
+		catchupErr = nil
+	}
+
+	// catchupCacheUpdate updates the catchupCache if this eventMsg seems to be the latest of its kind.
+	catchupCacheUpdate := func(ev *eventMsg) {
+		ts, ok := catchupCache[ev.event.Name]
+		if !ok || ev.apiTime.After(ts) {
+			catchupCache[ev.event.Name] = ev.apiTime
+		}
+	}
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing down main worker as context is finished", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.catchupPhaseRequest:
+			if catchupEventCh != nil {
+				client.Logger.Warn("Switching to catch-up-phase was requested while already catching up, restarting phase")
+
+				// Drain the old catch-up-phase producer channel until it is closed as its context will be canceled.
+				go func(catchupEventCh chan *catchupEventMsg) {
+					for _, ok := <-catchupEventCh; ok; {
+					}
+				}(catchupEventCh)
+				catchupCancel()
+			}
+
+			client.Logger.Info("Worker enters catch-up-phase, start caching up on Event Stream events")
+			catchupReset()
+			catchupEventCh, catchupCancel = client.startCatchupWorkers()
+
+		case catchupMsg, ok := <-catchupEventCh:
+			// Process an incoming event
+			if ok && catchupMsg.error == nil {
+				client.CallbackFn(catchupMsg.eventMsg.event)
+				catchupCacheUpdate(catchupMsg.eventMsg)
+				break
+			}
+
+			// Store an incoming error as the catchupErr to be processed below
+			if ok && catchupMsg.error != nil {
+				catchupErr = catchupMsg.error
+				break
+			}
+
+			// The channel is closed, replay cache and eventually switch modes
+			if len(catchupBuffer) > 0 {
+				// To not block the select and all channels too long, only one event will be processed per iteration.
+				ev := catchupBuffer[0]
+				catchupBuffer = catchupBuffer[1:]
+
+				ts, ok := catchupCache[ev.Name]
+				if !ok {
+					client.Logger.Debugw("Event to be replayed is not in cache", zap.Stringer("event", ev))
+				} else if ev.Time.Before(ts) {
+					client.Logger.Debugw("Skip replaying outdated Event Stream event", zap.Stringer("event", ev),
+						zap.Time("event timestamp", ev.Time), zap.Time("cache timestamp", ts))
+					break
+				}
+
+				client.CallbackFn(ev)
+				break
+			}
+
+			if catchupErr != nil {
+				client.Logger.Warnw("Worker leaves catch-up-phase with an error, another attempt will be made", zap.Error(catchupErr))
+				go func() {
+					select {
+					case <-client.Ctx.Done():
+					case client.catchupPhaseRequest <- struct{}{}:
+					}
+				}()
+			} else {
+				client.Logger.Info("Worker leaves catch-up-phase, returning to normal operation")
+			}
+
+			catchupReset()
```

Comment:
If I understand the code correctly, if an error occurs, it will still continue processing `catchupEventCh` until it's closed, then, process all buffered events from the event stream and then handle the error. Did I get this correctly and if so, why is it done this way? I think I'd have treated all errors pretty much the same: stop everything and start from the beginning, like it should also happen if Icinga 2 was restarted.

Reply:
Currently, the only case when an error will be returned to the `catchupEventCh` is after the `startCatchupWorker`'s `errgroup` has finished with an error which is not a `context.Canceled`. Directly following, the `catchupEventCh` will be closed.

https://github.com/Icinga/icinga-notifications/blob/d6c4d36f07a3a27e6bbd62c522d4aef3de6dd2ad/internal/icinga2/client.go#L228-L238

Thus, next time the `case catchupMsg, ok := <-catchupEventCh:` will be entered, it is because `catchupEventCh` is closed and all `catchupBuffer` events will be processed. Please note, those events are not from the catch-up phase, but received from the Event Stream API in the meantime.

However, you have a point that there is no use in replaying those when the catch-up failed. By restarting the catch-up phase in case of an error, those "new" events will be queried from the Icinga 2 API again.

I am going to rework this. You might wanna ignore my response, it's mostly here to collect my thoughts.